### ISSUE TITLE
Support recuring events

### ISSUE THREAD

**Issue by C4ptainCrunch** [ics-py/ics-py#14]

**Comment by simonsmiley** [comment]
I am interested in using this library for my application, but need recurrence support. If you tell me what you want, I can even implement it myself

**Comment by guyzmo** [comment]
what's blocking from implementing this (besides having enough free time to code ;) )

About the comments from @Astalaseven, there are a few things that are fuzzy:

>    e.make_recurrent(each_week) # or with rfc standard ?
> type(e) should become a new type, for example a REvent pointing to a (event)list.

afaict, it's not possible to promote an instance to a new type with a method call. You could make `e.make_recurrent()` return a `REvent`, but that would make the API confusing (and not very pythonic). All in all I find @Astalaseven design suggestion to be slightly over-engineered and hard to implement.

I think it's better to keep it simple, and have a "recurring" property on the event, set as its own instance:

    e = Event(begin=x, end=y)
    e.set_recurrent(timedelta(weeks=1))
    assert e.recurrence != None
    assert e.recurrence.length == timedelta(weeks=1)

Then to set a limited number of recurrences:

    e = Event(begin=x, end=y)
    e.set_recurrent(timedelta(weeks=1), count=100)
    assert e.recurrence.count == 100

And to handle "exceptions" of the recurrence we could add:

    e.recurrence.add_exception(datetime(…))

which would remove all recurrences from the period defined by datetime (or throw an exception if the period matches no recurrences).

Then the `.begin` and `.end` members set up to the "first" event, to be as close to the RFC as we can, but we could have a generator to create event objects for each recurring event:

    for rec_event in e.recurrence.items():
        print(rec_event)

as a real world example we could lookup whether a given time point is free:

    def check_timepoint_is_free(timepoint):
        for event in cal.events:
            if event.recurrence:
                for re in event.recurrence.items():
                    if time_point > re.begin and time_point < re.end:
                        return False
            else:
                if time_point > re.begin and time_point < re.end:
                    return False
        return True

then that would be up to discussion to consider how to implement those generated events:

- whether we create a `RecEvent` object that would have no `uid` but an event property that points to the first defining event ;
- or we use a `Event` objects that share the same `uid` as the first event.

Here the implementation would be quite easy:

    class Recurrence:
        def __init__(self, recurrence=None, count=float('inf'), exceptions=set()):
            self.recurrence = recurrence
            self.count = count
            self.exceptions = exceptions

        def add_exception(self, exception):
            # make sure that the exception is valid!
            self.exceptions.add(exception)

        def items(self):
            idx = 0
            while idx < self.count:
                idx += 1
                event = Event.clone() # or do: event = RecEvent(event, shift=self.recurrence*idx)
                event.begin = event.begin + (self.recurrence*idx)
                event.end = event.end + (self.recurrence*idx)
                yield event

    class Event:
        def set_recurrence(self, recurrence=None, count=None, exceptions=[]):
            self.recurrence = Recurrence(recurrence, count, exceptions)

And of course, we should then update all the code that generates RFC5545.

The advantages of the solution I'm suggesting are:

 - it's easy to write tests for that ;
 - it's a simple implementation, changing little of the existing code ;
 - the API is imho quite pythonic and easy to use.

**Comment by guyzmo** [comment]
ok, I started an implementation (still too dirty to be pushed anywhere), and actually the really painful thing is the `BY*` rules!

**Comment by GMLudo** [comment]
For now, I need to use a C middleware to use libical to have recurrent events support in my Python application.
Good to know that somebody works on that :-)

**Comment by C4ptainCrunch** [comment]
> what's blocking from implementing this (besides having enough free time to code ;) )

Time, time and time :/ (and a brain to figure the best api)

@guyzmo your API seems quite simple and pythonic :)

We still need to find a way to show those events in the `timeline`

**Comment by guyzmo** [comment]
Yup, I actually wrote that as I was playing with the "stable pypi" version of the module, then when I started working on the other PR, I found out about the timeline API. Though, a good API for recurrence events should make it not too hard to keep the timeline API clean.

The only thing will be to keep it efficient.

**Comment by encukou** [comment]
The dateutil package has pretty good support for [recurrence rules](https://dateutil.readthedocs.io/en/stable/rrule.html). Would it be possible to "just" add a property to Event that would contain a dateutil rruleset?

**Comment by danizen** [comment]
So, I found a way around this, for those who just need to read recurrences, rather than create them:

```python
def rrule(self):
    for content_line in self._unused:
        if content_line.name == 'RRULE':
            return content_line.value
```

This really demonstrates to me why the feature is hard - you need to represent the recurrence, parse the recurrence, and provide a Pythonic API to create the recurrence, all while keeping other APIs working.

While a simple property like the above works for my use case, it may not work for truly big files.

**Comment by danizen** [comment]
This is very complicated - https://icalendar.org/iCalendar-RFC-5545/3-8-5-3-recurrence-rule.html
In my case, I built a daily appointment and then canceled one of them to see how specifically Exchange (my version) would export it.

Here is is what I ended up with in event._unused:

```python
    In [9]: list(events[0]._unused)
    Out[9]:
    [<ContentLine 'CLASS' with 0 parameter. Value='PUBLIC'>,
     <ContentLine 'EXDATE' with 1 parameter. Value='20190314T083000'>,
     <ContentLine 'RRULE' with 0 parameter. Value='FREQ=WEEKLY;COUNT=10;BYDAY=MO,TU,WE,TH,FR'>,
     <ContentLine 'SEQUENCE' with 0 parameter. Value='0'>,
     <ContentLine 'X-MICROSOFT-CDO-BUSYSTATUS' with 0 parameter. Value='BUSY'>]
```

So, for my implementation, I will only need to consider RRULE and EXDATE until something more complicated shows up.   But, I'll need pretty good support for those to schedule dental appointments up to 90 days out in light of my work schedule.

The trick to calculating the recurrence can be seen by considering this example from the above:

```
DTSTART;TZID=America/New_York:19970902T090000
EXDATE;TZID=America/New_York:19970902T090000
RRULE:FREQ=MONTHLY;BYDAY=FR;BYMONTHDAY=13
```

BYDAY selects every Friday, BYMONTHDAY filters to the 13th when it occurs on a Friday.  So, all the BY must apply equally.  Then, you take away the EXDATE instances.

I'm actually never in the office by 8:30 am, and that's how I know it is a fake event, for now.

**Comment by danizen** [comment]
Anybody looking at this, do not be confused - I am not going to fix this issue, but I may provide some test data eventually.   At least I am providing here some insight on how RRULE works in practice.

As I moved and rescheduled more instances of this fictitious appointment, I discevered that EXDATE will accumulate additional values:

```python
In [9]: list(cal.events[0]._unused)
Out[9]:
[<ContentLine 'CLASS' with 0 parameter. Value='PUBLIC'>,
 <ContentLine 'EXDATE' with 1 parameter. Value='20190314T083000,20190321T083000,20190322T083000'>,
 <ContentLine 'RDATE' with 1 parameter. Value='20190322T100000'>,
 <ContentLine 'RRULE' with 0 parameter. Value='FREQ=WEEKLY;COUNT=10;BYDAY=MO,TU,WE,TH,FR'>,
 <ContentLine 'SEQUENCE' with 0 parameter. Value='0'>,
 <ContentLine 'X-MICROSOFT-CDO-BUSYSTATUS' with 0 parameter. Value='BUSY'>]
```

Hope this helps.

**Comment by danizen** [comment]
Blog post that explains a bit more:
https://www.nylas.com/blog/calendar-events-rrules/

I also noticed that Daylight standard  time is encoded at the start of my calendar export as follows:

```
BEGIN:VTIMEZONE
TZID:Eastern Standard Time
BEGIN:STANDARD
DTSTART:16011104T020000
RRULE:FREQ=YEARLY;BYDAY=1SU;BYMONTH=11
TZOFFSETFROM:-0400
TZOFFSETTO:-0500
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:16010311T020000
RRULE:FREQ=YEARLY;BYDAY=2SU;BYMONTH=3
TZOFFSETFROM:-0500
TZOFFSETTO:-0400
END:DAYLIGHT
END:VTIMEZONE
```

**Comment by Riamse** [comment]
Are there any updates on this? It's a 6 year old issue for a 20+ year old standard. Is it possible to use some other library in conjunction with ics.py to fill in this gap somehow?

**Comment by GMLudo** [comment]
To my knowledge, no Python library has this support.
However, you can use [ical.js](https://github.com/mozilla-comm/ical.js/), put it in a small NodeJS daemon and interact with it.

**Comment by aceisace** [comment]
Just adding on top of this issue as one of my projects requires the correct parsing of recurring events. Not sure how pythonic this is or if it can be implemented in ics, although I really wish it could.

According to my tests, rrules can be categorised in 3 categories:
* Minimal example of a recurring event using 'FREQ' (string):
```python
BEGIN:VEVENT
RRULE:FREQ=WEEKLY;COUNT=3;INTERVAL=1;BYDAY=FR
DTSTAMP:20190212T202300Z
DTSTART;VALUE=DATE:20190111
DTEND:20190112T000000Z
SUMMARY: some event
END:VEVENT
```
* Minimal example of a recurring event using 'UNTIL' with timezone data(string):
```python
BEGIN:VEVENT
RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20200330T215959Z;BYDAY=MO
DTSTAMP:20191214T151800Z
DTSTART:20191209T110700Z
DTEND:20191209T120700Z
SUMMARY: some recurring event
END:VEVENT
```
* Minimal example of a recurring event with until, but no timezone:
```python
BEGIN:VEVENT
RRULE:FREQ=WEEKLY;UNTIL=20190321;INTERVAL=2
DTSTAMP:20191214T151800Z
DTSTART;VALUE=DATE:20190307
DTEND:20190308T000000Z
SUMMARY:Recurring Event
END:VEVENT
```

I came up with this method to parse the rrules:

```python
from ics import Calendar
import arrow

import re
from dateutil.rrule import rrulestr

ical = str(urlopen(  URL ).read().decode())   # get iCalendar from URL

for events in ical:                                          # Search all events in the Calendar
  if re.search('RRULE:(.+?)\n', str(events)):     # Check if RRULE is present in events
    rule = re.search('RRULE:(.+?)\n', event_str).group(0)[:-2]                # Extract RRULE
    # FREQ and UNTIL with timezone (example 1 & 2) can be parsed directly with the rrule.
    # If RRULE contains UNTIL but no timezone, an Error is thrown.
    # For those rrules, the timezone has to be added manually:
    if re.search('UNTIL=(.+?);', rule) and not re.search('UNTIL=(.+?)Z;', rule):
      rule = re.sub('UNTIL=(.+?);', 'UNTIL='+re.search('UNTIL=(.+?);', rule).group(0)[6:-1]+'T000000Z;', rule)

    dates = rrulestr(rule, dtstart= events.begin.datetime)         # This gets the dates of all recurrences

    # With the dates of the recurrences, events can be cloned and modified to the recurrence date
    if dates:                                                    # Check if the list of dates exists (is not empty)
      duration = events.duration                    # Get the duration of the event
      for date in dates:                                    # For each date
        cc = events.clone()                               # Clone the original event
        cc.end = arrow.get(date+duration)     # Replace the end of cloned event to the date + duration of the recurrence
        cc.begin = arrow.get(date)                  # Replace the beginning time of the cloned event to the date of recurrence

```
After that, the cloned event with the modified date can maybe be added to the iCalendar. Or at least, that is what I would do.

**Comment by clstaudt** [comment]
Please implement support for recurring events. 
-- maintainer of a project based on ics-py

**Comment by DavidZiemann** [comment]
Based on the work [aceisace](https://github.com/aceisace) did in his comment above, I tweaked the logic to work for my specific situation & hopefully it helps someone out. I am working on a script to retrieve my calendar for a specific date range. This logic helped me limit the scope of the dates returned to improve the performance of the method since the dates list can be masive based on your `rrule`.

```Python
import datetime
import arrow
import re

from dateutil.rrule import rrulestr
from ics import Event

def generate_recurring_events(cal, start_date, end_date):
  events_to_add = []
  for event in cal.events:
    if re.search('RRULE:(.+?)\n', str(event)):
      rule = re.search('RRULE:(.+?)\n', str(event)).group(0)[:-2]
      if re.search('UNTIL=(.+?);', rule) and not re.search('UNTIL=(.+?)Z;', rule):
        rule = re.sub('UNTIL=(.+?);', 'UNTIL='+re.search('UNTIL=(.+?);', rule).group(0)[6:-1]+'T000000Z;', rule)
      
      # Only retrieve the dates within the requested timeframe
      dates = list(rrulestr(rule, dtstart= event.begin.datetime).between(start_date, end_date, True))

      if dates:
        duration = event.duration
        for date in dates:
          # You can clone the event if you want to preserve all details
          e = Event()
          e.name = event.name
          e.end = arrow.get(date+duration)
          e.begin = arrow.get(date)
          events_to_add.append(e)

  for event in events_to_add:
    cal.events.add(event)
```

**Comment by allenporter** [comment]
I came across a nice discussion of a design for recurring events: https://github.com/bmoeskau/Extensible/blob/master/recurrence-overview.md

**Comment by Kylmakalle** [comment]
You probably want to handle EXDATE for some cancelled/deleted occurrences of recurring events.


This is a code you can inject before `dates = list(rrulestr(...))` line based on @DavidZiemann's example

```python
exdate_search = re.search('EXDATE(.+?)\n', str(event))
if exdate_search:
    exdate = exdate_search.group(0)[:-2]
    rule += f"\n{exdate}"
```

**Comment by Kylmakalle** [comment]
I also struggled with my outlook calendar. For some reason there's issue with timezones for me. Even with `T000000` some events says the're in my local Timezone, not UTC. So here's my patch to that.
I hope you don't need it anyways.

```python
if 'TZID=' in exdate:
    exdate = re.sub('TZID=([^:]+):', 'TZID=UTC:', exdate)
```

**Comment by allenporter** [comment]
I don't think that is great advice given TZID of UTC is not valid https://www.rfc-editor.org/rfc/rfc5545#section-3.2.19. 


For discussing actual real solutions for this problem: I've implemented recurrence rules in another python library following the design linked above
[ical/timeline.py](https://github.com/allenporter/ical/blob/c4381addd763194875190b11a1be455756b831c8/ical/timeline.py#L172) which puts recurring events behind the `Timeline` modeled after ics-py. It uses `dateutil.rrule` as the underlying library to expand the recurring rules. The timeline underneath creates an iterator for normal events and an [iterator for each recurring event](https://github.com/allenporter/ical/blob/c4381addd763194875190b11a1be455756b831c8/ical/iter.py#L26), then provides a [merged iterator](https://github.com/allenporter/ical/blob/c4381addd763194875190b11a1be455756b831c8/ical/iter.py#L98) that picks the next event from each iterator. Something like this can work for `ics-py` also since basically since it is modeled after a `Timeline`.

**Comment by Kylmakalle** [comment]
> given TZID of UTC is not valid https://www.rfc-editor.org/rfc/rfc5545#section-3.2.19

>> Note: This document does not define a naming convention for
         time zone identifiers
         
I assume that `UTC` is a pretty common (and also valid for dateutil's `rrulestr`) timezone identifier

**Comment by allenporter** [comment]
> > given TZID of UTC is not valid https://www.rfc-editor.org/rfc/rfc5545#section-3.2.19
> 
> >> Note: This document does not define a naming convention for
>          time zone identifiers
>          
> I assume that `UTC` is a pretty common (and also valid for dateutil's `rrulestr`) timezone identifier

> The "TZID" property parameter MUST NOT be applied to DATE properties and DATE-TIME or TIME properties whose time values are specified in UTC.

I realize this can mean a time value ending in Z but I'm referring to the spirit of this. Anyway, hack whatever you want.

**Comment by danizen** [comment]
It is timely (see what I did there) that this comes up now.  I've moved to a new employer who gave me a Mac rather than a PC.  This complicates things, because instead of reading events from Exchange via MAPI, I now have to export a file from Outlook for Mac, and parse it, and there is limited ICS data in there.  Having become an knowledgeable on ICS and RRULE, I now no longer need it :).  At least for now...

**Comment by 0dminnimda** [comment]
There is a library that implements the recurrance, maybe we can borrow something from them?
https://pypi.org/project/recurring-ical-events/

**Comment by 0dminnimda** [comment]
Also, in case I will find the time to work on this, where should I start and is pr for this feature a welcome one?

**Comment by allenporter** [comment]
I would recommend you start with the ical library I linked since the interfaces are based on ics-py.

**Comment by N-Coder** [comment]
There's multiple facets to this topic. I guess I see the following points:

1. Representing & Interpreting the recurrence information: dateutil handles both, altought their representation is somewhat hard to use and might have issues with generating the recurrence’s, see their issues X and Y. Actually computing this according to the standard is very hard, so it’s still good to have it in a specialised library.
2. Listing Events and their recurrences in order: the Timeline class already handles most of it, its heap-based data structure should be very easy to adapt to also respect the recurrences generated by dateutil.
3. Actually representing recurrences of an event. Recurring events might have two forms. Either they appear as a simple repetition of the base event, i.e. they can be seen as reference to the base event + the time stamp of the actual recurrence as calculated by dateutil (and maybe a counter how often the event already occurred). In this case, all the occurrences share the properties of the base event. The second form becomes relevant when you start modifying single recurrences. This generates a second, independent event in the data file with potentially different properties. Via an indirection defined in the standard, it references the base event it stems from. The recurrence rule of the base event has an exception for the single recurrences that appears individually. How do we best represent these two types of event instances next to each other? Should both have the same interface or should we use the triple (date, occurrence nr, event object) - or something different - also for the exception instances and normal events when listed in the Timeline? And also how do we link the exception instance with the base instance?
4. Which modification operations should be support? I’m thinking of adding exceptions and creating modifiable copies for single or multiple occurrences (given by a time range or another recurrence rule?), deleting exceptions and possibly also the linked copy, splitting recurring events in two independent halves at a certain date, … Any further ideas?

I guess points one and two are very easy to do while the other two still need some thinking. I’d first create and discuss a draft for the (read-only) interface of point three, implementing points one and two in a PoC for that should then been easy. Coming up for a WIP PR here would be the next thing on my todo list after the 0.8 release. The functionality for modification can probably then be done in a second go, but creating a GH Discussion and already collecting ideas might be a good start for that. A maybe fifth point would coming up with tests for all this, which can also be already started now. (And believe me, testing is currently the biggest issue I have with the next release - I can happily work on the remaining issues, but I have a really hard time building trust into the result. So if you want to contribute, working on the tests is easy to do and would help a lot. Here we could even go with test-driven design, i.e. having meaningful tests that describe/specify how things should work first. Other than that, porting and improving the test set is a really big point that would be easy to work on.)

As the previous discussion here is a pretty mixed bag of different points from above, we should maybe keep future discussion here on the structure of this "meta" issue and move discussion of individual points (e.g. those from my list) to individual issues/PoC PRs/GH discussions.

**Comment by allenporter** [comment]
To elaborate on my comment above... I implemented recurring events library following a model based on the ics-py timeline class, and it's working great. 

This is my suggestion from doing this:
1) Yes, use dateutil.ruleset. It can handle rrules and rdate.  [example](https://github.com/allenporter/ical/blob/aec74f333eb440b0cc26b18612fd0d1514f8cbbd/ical/timeline.py#L178)
2) Make a single iterator for all recurring events using a heap of iterators [example](https://github.com/allenporter/ical/blob/aec74f333eb440b0cc26b18612fd0d1514f8cbbd/ical/iter.py#L98)
3) Return each instance as a separate event with a different  start/end time.  Keep the same `uid` but update the `recurrence_id`.  [example](https://github.com/allenporter/ical/blob/aec74f333eb440b0cc26b18612fd0d1514f8cbbd/ical/timeline.py#L167)
4) Edit operations take a `uid`, `recurrence_id` (optional), and `recurrence_range` which means either "this event" or "this event and future events" following the suggestions in rfc 4465.

**Comment by make-github-pseudonymous-again** [comment]
> 2\. Make a single iterator for all recurring events using a heap of iterators

The only problem I see with this approach is that it does not allow to query a time interval efficiently. But it can be a good enough starting solution. We can see later what most common performance problems people hit with this solution (if at all).

**Comment by allenporter** [comment]
Yes, that's right. My impression is this is true of timeline already, though it is exaggerated in this approach.

**Comment by C4ptainCrunch** [comment]
From [#315](https://github.com/ics-py/ics-py/discussions/315#discussioncomment-4406027)
> Yes, I ended up looking to extend ics-py for home assistant use cases but ended up with some slightly different design goals so started a new project. The timeline and timespan APIs definitely straight modeled after ics-py.

ical is [supporting recurring-events](https://allenporter.github.io/ical/ical.html#recurring-events) so i guess we could have a look at their implementation to get inspiration for ours (that would be a funny way of closing the loop 😁 )

**Comment by allenporter** [comment]
Yeah I posted the links to the approach I took in ical on the ics-py discussion for this. Happy to share back code there.

**Comment by mwunderlich** [comment]
Related to this: It is currently possible to add the recurrence rule to the event object like this: 
`event.extra.append("RRULE:FREQ=WEEKLY")`

However, when trying to write the calendar to a file, this will lead to the following error: 

```
  File "[...]/lib/python3.11/site-packages/ics/grammar/parse.py", line 146, in clone
    c.append(elem.clone())
AttributeError: 'str' object has no attribute 'clone'
```

**Comment by mwunderlich** [comment]
> Related to this: It is currently possible to add the recurrence rule to the event object like this: `event.extra.append("RRULE:FREQ=WEEKLY")`
> 
> However, when trying to write the calendar to a file, this will lead to the following error:
> 
> ```
>   File "[...]/lib/python3.11/site-packages/ics/grammar/parse.py", line 146, in clone
>     c.append(elem.clone())
> AttributeError: 'str' object has no attribute 'clone'
> ```

PS: As a work around to this, it is always possible to export the Calendar object to file and modify this output file as a text file to add the recurrence. E.g. the following will add a weekly recurrence to every event in the calendar: 
```

    # Add weekly recurrence rule to each event by modifying the ICS file as text
    with open(outfile, 'r') as ics_file:
        lines = ics_file.readlines()

    with open(outfile, 'w') as ics_file:
        for line in lines:
            ics_file.write(line)
            # Add recurrence rule after the event's end time
            if line.startswith("DTEND:"):  # Insert the RRULE after the end time of each event
                ics_file.write("RRULE:FREQ=WEEKLY\n")
```
